File: acoustic/vad_gate.py
"""
vad_gate.py
------------------
Owner: Person A (acoustic stream)

Wraps the official Silero VAD ONNX model (snakers4/silero-vad, MIT license)
to gate audio before it reaches AASIST-L and the semantic stream (Person B's
Zipformer). This is the "NEW" component from the architecture update - it
cuts wasted compute on silence/hold-music by only letting speech segments
through to the two heavier models downstream.

Confirmed I/O contract by loading the actual model (not documentation -
the model file itself):
  inputs:  "input" [1, N] float32 audio chunk
           "state"  [2, 1, 128] float32 - recurrent state, carried between
                     calls (this is the detail people most often get wrong -
                     forgetting to feed the updated state back in makes the
                     VAD much less accurate)
           "sr"     scalar int64 sample rate (16000)
  outputs: "output"  [1, 1] float32 speech probability
           "stateN"  [2, 1, 128] float32 - new state, feed into next call

Standard chunk size for this model is 512 samples (32ms) at 16kHz - feed it
consistently at that size for best results, matching how it was trained.
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import onnxruntime as ort

_THIS_DIR = Path(__file__).parent
_MODEL_PATH = _THIS_DIR / "aasist_model" / "silero_vad.onnx"

SAMPLE_RATE = 16000
CHUNK_SAMPLES = 512  # 32ms at 16kHz - the size this model expects


class VadGate:
    def __init__(self, model_path: Path = _MODEL_PATH, threshold: float = 0.5):
        self.session = ort.InferenceSession(str(model_path), providers=["CPUExecutionProvider"])
        self.threshold = threshold
        self._state = np.zeros((2, 1, 128), dtype=np.float32)
        self._sr = np.array(SAMPLE_RATE, dtype=np.int64)

    def reset(self):
        """Call this at the start of each new call/session - stale recurrent
        state from a previous call will skew the first few predictions."""
        self._state = np.zeros((2, 1, 128), dtype=np.float32)

    def is_speech(self, chunk: np.ndarray) -> dict:
        """chunk: 1-D float32 array, ideally exactly CHUNK_SAMPLES long.
        Returns the speech probability and a boolean gate decision."""
        if chunk.shape[0] != CHUNK_SAMPLES:
            # Pad or trim rather than erroring - live audio callbacks don't
            # always hand you exactly 512 samples.
            if chunk.shape[0] < CHUNK_SAMPLES:
                chunk = np.pad(chunk, (0, CHUNK_SAMPLES - chunk.shape[0]))
            else:
                chunk = chunk[:CHUNK_SAMPLES]

        x = chunk.astype(np.float32)[np.newaxis, :]
        prob, new_state = self.session.run(
            None, {"input": x, "state": self._state, "sr": self._sr}
        )
        self._state = new_state  # carry state forward - see docstring above

        p = float(prob[0][0])
        return {"speech_prob": round(p, 4), "is_speech": p >= self.threshold}


# ---------------------------------------------------------------------------
# Self-test - loads the REAL Silero VAD model.
# Run directly: python3 vad_gate.py
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    print("Loading Silero VAD (real ONNX model, not a stub)...")
    vad = VadGate()

    rng = np.random.default_rng(0)
    silence = np.zeros(CHUNK_SAMPLES, dtype=np.float32)
    noise = (rng.standard_normal(CHUNK_SAMPLES) * 0.3).astype(np.float32)

    print("\n--- Test: silence chunk ---")
    print(vad.is_speech(silence))

    print("\n--- Test: loud noise chunk (not real speech, just non-trivial signal) ---")
    print(vad.is_speech(noise))

    print("\nNOTE: like the AASIST-L test, this only confirms the pipeline runs "
          "end-to-end with the real model - test against actual recorded "
          "speech/silence on Day 1, not synthetic noise.")

File: acoustic/android/VadGate.kt
/**
 * VadGate.kt
 * Owner: Person A (acoustic stream)
 *
 * Android port of vad_gate.py - same model (silero_vad.onnx), same I/O
 * contract, confirmed against the actual model file:
 *   inputs:  "input" [1, N] float32, "state" [2,1,128] float32,
 *            "sr" scalar int64
 *   outputs: "output" [1,1] float32 speech probability, "stateN" new state
 *
 * The recurrent state MUST be carried between calls (assign the returned
 * stateN back into the next call's state input) - this is the detail
 * that's easy to miss and silently degrades VAD accuracy if skipped.
 *
 * Setup: bundle silero_vad.onnx (from aasist_model/) as an Android asset,
 * add the ONNX Runtime Mobile dependency:
 *   implementation("com.microsoft.onnxruntime:onnxruntime-android:latest.release")
 */

package com.echoguard.acoustic

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import java.nio.FloatBuffer
import java.nio.LongBuffer

class VadGate(
    assetManager: android.content.res.AssetManager,
    modelAssetPath: String = "models/silero_vad.onnx",
    private val threshold: Float = 0.5f,
) {
    companion object {
        const val SAMPLE_RATE = 16000L
        const val CHUNK_SAMPLES = 512  // 32ms @ 16kHz - match this model's training chunk size
    }

    private val env = OrtEnvironment.getEnvironment()
    private val session: OrtSession
    private var state = FloatArray(2 * 1 * 128)  // zeroed recurrent state

    init {
        val modelBytes = assetManager.open(modelAssetPath).readBytes()
        session = env.createSession(modelBytes)
    }

    /** Call at the start of each new call/session - stale state from a
     * previous call skews the first few predictions. */
    fun reset() {
        state = FloatArray(2 * 1 * 128)
    }

    data class Result(val speechProb: Float, val isSpeech: Boolean)

    /** samples should be exactly CHUNK_SAMPLES (512) floats, 16kHz mono. */
    fun isSpeech(samples: FloatArray): Result {
        val padded = if (samples.size == CHUNK_SAMPLES) samples
                      else samples.copyOf(CHUNK_SAMPLES)  // zero-pads or truncates

        val inputTensor = OnnxTensor.createTensor(
            env, FloatBuffer.wrap(padded), longArrayOf(1, CHUNK_SAMPLES.toLong())
        )
        val stateTensor = OnnxTensor.createTensor(
            env, FloatBuffer.wrap(state), longArrayOf(2, 1, 128)
        )
        val srTensor = OnnxTensor.createTensor(
            env, LongBuffer.wrap(longArrayOf(SAMPLE_RATE)), longArrayOf()
        )

        val inputs = mapOf("input" to inputTensor, "state" to stateTensor, "sr" to srTensor)
        session.run(inputs).use { results ->
            val prob = (results.get(0).value as Array<FloatArray>)[0][0]
            val newState = results.get(1).value as Array<Array<FloatArray>>
            // Flatten [2,1,128] back into our stored FloatArray for the next call.
            state = FloatArray(2 * 128) { i -> newState[i / 128][0][i % 128] }
            return Result(speechProb = prob, isSpeech = prob >= threshold)
        }
    }

    fun release() {
        session.close()
    }
}
